diff --git a/src/nsterm.m b/src/nsterm.m
index d99c6e9875c..ab2f535d34d 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -9616,7 +9616,7 @@ - (instancetype) initWithEmacsFrame: (struct frame *) f
     styleMask = NSWindowStyleMaskBorderless;
   else if (FRAME_UNDECORATED (f))
     {
-      styleMask = NSWindowStyleMaskBorderless;
+      styleMask = NSWindowStyleMaskTitled;
 #ifdef NS_IMPL_COCOA
       styleMask |= NSWindowStyleMaskResizable;
 #endif
@@ -9834,7 +9897,7 @@ - (void)setParentChildRelationships
      drop-shadows break this effect, so remove them on undecorated
      child frames.  */
   if (parentFrame && FRAME_UNDECORATED (ourFrame))
-    [self setHasShadow:NO];
+    [self setHasShadow:YES];
   else
     [self setHasShadow:YES];
 #endif
